pkg/registration/spoke/registration/hub_timeout_controller.go

```diff
@@ -0,0 +1,70 @@
+package registration
+
+import (
+	"context"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	coordlisters "k8s.io/client-go/listers/coordination/v1"
+	"k8s.io/klog/v2"
+)
+
+const leaseName = "managed-cluster-lease"
+
+type hubTimeoutController struct {
+	clusterName        string
+	leaseLister        coordlisters.LeaseLister
+	timeoutSeconds     int32
+	lastLeaseRenewTime time.Time
+	handleTimeout      func(ctx context.Context) error
+}
+
+func NewHubTimeoutController(
+	clusterName string,
+	leaseLister coordlisters.LeaseLister,
+	timeoutSeconds int32,
+	handleTimeout func(ctx context.Context) error,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &hubTimeoutController{
+		clusterName:    clusterName,
+		timeoutSeconds: timeoutSeconds,
+		handleTimeout:  handleTimeout,
+		leaseLister:    leaseLister,
+	}
+	return factory.New().WithSync(c.sync).ResyncEvery(time.Minute).
+		ToController("HubTimeoutController", recorder)
+}
+
+func (c *hubTimeoutController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	logger := klog.FromContext(ctx)
+	if c.handleTimeout == nil {
+		return nil
+	}
+
+	lease, err := c.leaseLister.Leases(c.clusterName).Get(leaseName)
+	if err != nil {
+		logger.Error(err, "Failed to get lease", "cluster", c.clusterName, "lease", leaseName)
+		// This means to handle the case which the hub is not connectable from the beginning.
+		if c.lastLeaseRenewTime.IsZero() {
+			c.lastLeaseRenewTime = time.Now()
+		}
+	} else {
+		c.lastLeaseRenewTime = lease.Spec.RenewTime.Time
+	}
+
+	if isTimeout(time.Now(), c.lastLeaseRenewTime, c.timeoutSeconds) {
+		logger.Info("Lease timeout", "cluster", c.clusterName, "lease", leaseName)
+		err := c.handleTimeout(ctx)
```

Comment:
I think I understand what you are trying to do...So the you get from local cache, and if the network is disconnected, the local cache will not be updated? If so I would rather use client to get and maintain in this controller, using informer and lister is too heavy.

Also we should add a TODO. These code should finally be put into the existing lease controller.

Reply:
Sure, that makes sense.